langchain/document_loaders/git.py

```diff
@@ -0,0 +1,72 @@
+import os
+from typing import List, Optional
+
+from langchain.docstore.document import Document
+from langchain.document_loaders.base import BaseLoader
+
+
+class GitLoader(BaseLoader):
+    """Loads files from a local Git repository into a list of documents.
+    Currently supports only text files.
+
+    Each document represents one file in the repository. The `path` points to
+    the local Git repository, and the `branch` specifies the branch to load
+    files from. By default, it loads from the `main` branch.
+    """
+
+    def __init__(
+        self,
+        path: str,
+        branch: Optional[str] = "main",
+    ):
+        self.path = path
+        self.branch = branch
+
+    def load(self) -> List[Document]:
+        try:
+            from git import Blob, Repo
+        except ImportError as ex:
+            raise ImportError(
+                "Could not import git python package. "
+                "Please install it with `pip install GitPython`."
+            ) from ex
+
+        repo = Repo(self.path)
+        repo.git.checkout(self.branch)
+
+        docs: List[Document] = []
+
+        for item in repo.tree().traverse():
+            if isinstance(item, Blob):
+                file_path = os.path.join(self.path, item.path)
+                rel_file_path = os.path.relpath(file_path, self.path)
+                try:
+                    with open(file_path, "rb") as f:
+                        content = f.read()
+                        file_type = os.path.splitext(item.name)[1]
+
+                        # loads only text files
+                        if self.is_text_content(content):
+                            metadata = {
+                                "file_path": rel_file_path,
+                                "file_name": item.name,
+                                "file_type": file_type,
+                            }
+                            text_content = content.decode("utf-8", errors="ignore")
+                            doc = Document(page_content=text_content, metadata=metadata)
+                        else:
+                            continue
+                        docs.append(doc)
+                except Exception as e:
+                    print(f"Error reading file {file_path}: {e}")
+
+        return docs
+
+    @staticmethod
+    def is_text_content(content: bytes) -> bool:
+        """Determines if the content is text based on the content bytes."""
+        try:
+            content.decode("utf-8")
+            return True
+        except UnicodeDecodeError:
+            return False
```

Comment:
maybe return the `content.decode("utf-8")` so that we don't have to decode twice

Reply:
done https://github.com/hwchase17/langchain/pull/2851/commits/3a16d351cccf8b36036a046cc86606e11fd9b084